File: guwave-onetest/src/main/java/com/guwave/onetest/site/helper/domain/Entity.java
package com.guwave.onetest.site.helper.domain;

import java.util.Objects;

public abstract class Entity<IDType> {
    private final IDType id;

    public Entity(IDType id) {
        if (Objects.isNull(id)) {
            throw new IllegalArgumentException("entity id can not be null");
        }
        this.id = id;
    }

    public IDType id() {
        return this.id;
    }

    public boolean sameIdentityAs(Entity<?> other) {
        if (other == null) {
            return false;
        }

        return Objects.equals(id, other.id);
    }
}

File: GBDemo/PinManagement/Helper/Domain/Entity.cs
using System;
using System.Diagnostics.CodeAnalysis;

namespace PinManagement.Helper.Domain
{
    public abstract class Entity<IDType> : IEquatable<Entity<IDType>>
    {
        private readonly IDType id;

        protected Entity(IDType id)
        {

            this.id = id;
        }

        public IDType Id
        {
            get { return this.id; }
        }

        public override bool Equals(object obj)
        {
            var entity = obj as Entity<IDType>;
            if (entity != null)
            {
                return this.Equals(entity);
            }
            return base.Equals(obj);
        }

        public override int GetHashCode()
        {
            return this.Id.GetHashCode();
        }

        public bool Equals(Entity<IDType> other)
        {
            if (other == null)
            {
                return false;
            }
            return this.Id.Equals(other.Id);
        }
    }
}
